app/contexts/AppStateContext.tsx

```diff
@@ -0,0 +1,85 @@
+import React, { createContext, useCallback, useContext, useEffect, useRef } from 'react'
+import { AppState, AppStateStatus } from 'react-native'
+
+export type SimplifiedAppStateStatus = 'active' | 'background'
+export type AppStateListener = (status: SimplifiedAppStateStatus) => void
+
+export interface AppStateContext {
+  addListener: (listener: AppStateListener) => number
+  removeListener: (id: number) => void
+}
+
+const appStateContext = createContext<AppStateContext>(undefined as any)
+
+export function useAppStateContext (): AppStateContext {
+  return useContext(appStateContext)
+}
+
+export function AppStateContextProvider (props: React.PropsWithChildren<any>): JSX.Element | null {
+  // DO NOT use `useState`, these are meant for appstate monitoring
+  // re-render side effect is undesireable
+  const listenerRef = useRef<{ [key: number]: AppStateListener } | null>(null)
+  const prevState = useRef<AppStateStatus>('active') // first mounted
+
+  // all useCallback / useEffect in this context provider MUST have ZERO dependencies, to remain predictable
+  const emit = useCallback((listeners: AppStateListener[], nextState: SimplifiedAppStateStatus) => {
+    prevState.current = nextState // simplify iOS lifecyle, skipping `inactive` as transition state
+    listeners.forEach(l => l(nextState))
+  }, [])
+
+  const appStateHandler = useCallback((nextState: AppStateStatus) => {
+    const listeners = Object.values(listenerRef.current ?? {})
+    if (nextState === 'background' || nextState === 'inactive') {
+      if (prevState.current === 'active') {
+        emit(listeners, 'background')
+      }
+    } else if (nextState === 'active') {
+      if (prevState.current === 'background') {
+        emit(listeners, 'active')
+      }
+    }
+  }, [])
+
+  useEffect(() => {
+    const cb = (ns: AppStateStatus): void => {
+      appStateHandler(ns)
+    }
```

Comment:
```suggestion
    function resolveState(): string {
      ...
    }

    function onChange(status: AppStateStatus): void {
      prevState.current = status
      const state = resolvedState(...)
      
      for (listener ... ) {
        listener(state)
      }
    }
```

Should refactor and simplify this code, unnecessarily complex. From where I see, you just need to:

1. store prev state
2. resolve new state
3. for each emit out the newly resolved state